loopy/codegen/tools.py

```diff
@@ -0,0 +1,141 @@
+__copyright__ = "Copyright (C) 2020 Kaushik Kulkarni"
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+from pytools import ImmutableRecord, memoize_method
+from loopy.schedule import (EnterLoop, LeaveLoop, CallKernel, ReturnFromKernel,
+                            Barrier, BeginBlockItem, gather_schedule_block)
+
+
+class CodegenOperationCacheManager(ImmutableRecord):
+    def __init__(self, kernel):
+        super().__init__()
+        super().__setattr__("kernel", kernel)
+
+    def __setattr__(self, key, val):
+        if key == "kernel":
+            # overwriting self.kernel is not allowed, as cache will become
+            # invalidated
+            raise ValueError("Cannot update CodegenOperationCacheManager.kernel,"
+                             " create a new instance instead.")
+
+        super().__setattr__(key, val)
+
+    @memoize_method
+    def find_active_inames_at(self, sched_index):
```

Comment:
Recursion here is just bad. For kernels like in #178, this reaches recursion limit.

Reply:
Resolved by moving `cgen_cachemanager` into `CodeGenerationState`.